src/satisfia/util/helper.py

```diff
@@ -1,84 +1,81 @@
 #!/usr/bin/env python3
 
-class Interval():
-	def __init__(self, left, right=None):
-		if hasattr(left, "__len__"):
-			if right != None:
-				raise TypeError()
-			if len(left) != 2:
-				raise TypeError()
-			self._left = left[0]
-			self._right = left[1]
-		else:
-			self._left = left
-			self._right = left if (right == None) else right
+from typing import NamedTuple
+
+def Interval(left, right=None):
+    if hasattr(left, "__len__"):
+        assert right is None
+        return _Interval(*left)
+    else:
+        return _Interval(left,  left if right is None else right)
+
+class _Interval(NamedTuple):
+	left: float
+	right: float
 
 	def __str__(self):
-		if self._left > self._right:
+		if self.left > self.right:
 			return "<empty>"
-		return "[{}, {}]".format(self._left, self._right)
+		return "[{}, {}]".format(self.left, self.right)
 
 	def __repr__(self):
-		return "Interval({}, {})".format(self._left, self._right)
+		return "Interval({}, {})".format(self.left, self.right)
 
 	def __contains__(self, item):
-		return self._left <= item <= self._right
+		return self.left <= item <= self.right
 
 	def __le__(self, other):
-		return other[0] <= self._left <= self._right <= other[1]
+		return other[0] <= self.left <= self.right <= other[1]
 
 	def __eq__(self, other):
-		return (self._left == other[0]) and (self._right == other[1])
+		return (self.left == other[0]) and (self.right == other[1])
 
 	def __and__(self, other):
-		return Interval(max(self._left, other[0]), min(self._right, other[1]))
+		return _Interval(max(self.left, other[0]), min(self.right, other[1]))
 
-	def __getitem__(self, item):
-		if item == 0:
-			return self._left
-		elif item == 1:
-			return self._right
-		else:
-			raise IndexError()
+def interpolate(x, l, y):
+    # denoted	x : l : y in formulas
+    diff = y - x
+    return x + l * diff
 
-	def __hash__(self):
-		return hash((self._left, self._right))
+def interpolate2(x, l, y):
+    diff = y - x
+    return ( x + l[0] * diff, x + l[1] * diff )
 
-	def __len__(self):
-		return 2
+def interpolateInt(x, l, y):
+    # one argument is an interval, so everything becomes an interval
+    x, l, y = Interval(x), Interval(l), Interval(y)
+    return Interval(x[0] + l[0] * (y[0] - x[0]), x[1] + l[1] * (y[1] - x[1]))
 
-def interpolate(x, l, y):
-	# denoted	x : l : y in formulas
+def relativePosition2(x, z, y):
+    if (diff := (y-x)) == 0.:
```

Comment:
Interesting, I wasn't aware of the `:=` operator in Python. Good to know...

Reply:
It's pretty new: https://peps.python.org/pep-0572/